File: examples/php-app/src/Calculator.php
<?php

declare(strict_types=1);

namespace Example\PhpApp;

/**
 * Tiny library for the Rehearse PHP example.
 * The interesting part isn't the code — it's the CI pipeline that runs
 * PHPUnit + PHPStan across PHP 8.2 / 8.3 / 8.4 in parallel.
 */
final class Calculator
{
    public function add(int|float $a, int|float $b): int|float
    {
        return $a + $b;
    }

    public function divide(int|float $a, int|float $b): int|float
    {
        if ($b === 0 || $b === 0.0) {
            throw new \DivisionByZeroError('Cannot divide by zero');
        }
        return $a / $b;
    }

    /**
     * @param array<int, int|float> $values
     */
    public function average(array $values): float
    {
        if (count($values) === 0) {
            throw new \InvalidArgumentException('Cannot average an empty list');
        }
        return array_sum($values) / count($values);
    }
}

File: examples/dotnet-app/src/MyApp/Calculator.cs
namespace MyApp;

/// <summary>
/// Tiny library for the Rehearse dotnet-app example.
/// The interesting part is the CI pipeline that targets net8.0 + net9.0 + net10.0
/// in parallel via the runner's matrix scheduler.
/// </summary>
public sealed class Calculator
{
    /// <summary>Adds two numbers.</summary>
    public double Add(double a, double b) => a + b;

    /// <summary>Divides a by b. Throws on division by zero.</summary>
    public double Divide(double a, double b)
    {
        if (b == 0) throw new DivideByZeroException("Cannot divide by zero");
        return a / b;
    }

    /// <summary>Returns the arithmetic mean of values. Throws on empty input.</summary>
    public double Average(IReadOnlyCollection<double> values)
    {
        if (values.Count == 0) throw new ArgumentException("Cannot average an empty list", nameof(values));
        return values.Sum() / values.Count;
    }
}
